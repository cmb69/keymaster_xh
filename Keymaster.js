/**
 * The Keymaster class.
 *
 * @file
 *
 * @version   SVN: $Id$
 * @author    Christoph M. Becker <cmbecker69@gmx.de>
 * @copyright 2013 Christoph M. Becker <http://3-magi.net/>
 * @license   http://www.gnu.org/licenses/gpl.html GNU GPLv3
 */

/**
 * The Keymaster class.
 *
 * @constructor
 *
 */
Keymaster = function () {
    /**
     * The warning dialog.
     *
     * @type {HTMLDivElement}
     *
     * @private
     */
    this._warningDialog = null;

    /**
     * The warning text.
     *
     * @type {Text}
     *
     * @private
     */
    this._warningText = '';

    /**
     * The timer handle.
     *
     * @type {Number}
     *
     * @private
     */
    this._timer = null;

    this.createWarningDialog();
    this._timer = setInterval(this.bind(this.requestRemainingTime),
            Keymaster.config.pollInterval);
};

/**
 * The configuration.
 *
 * @type {Object}
 *
 * @static
 */
Keymaster.config = null;

/**
 * The localization
 *
 * @type {Object}
 *
 * @static
 */
Keymaster.l10n = null;

/**
 * Creates the warning dialog.
 *
 * @method
 *
 * @protected
 */
Keymaster.prototype.createWarningDialog = function () {
    var dialog, message, para, text, div, button;

    dialog = document.createElement("div");
    dialog.id = "keymaster";
    dialog.style.display = "none";
    message = document.createElement("div");
    message.className = "keymaster_message";
    para = document.createElement("p");
    text = document.createTextNode("");
    this._warningText = text;
    para.appendChild(text);
    message.appendChild(para);
    div = document.createElement("div");
    div.className = "keymaster_buttons";
    button = document.createElement("button");
    button.onclick = this.bind(this.resetSession);
    text = document.createTextNode(Keymaster.l10n.button);
    button.appendChild(text);
    div.appendChild(button);
    message.appendChild(div);
    dialog.appendChild(message);
    document.body.appendChild(dialog);
    this._warningDialog = dialog;
};

/**
 * Shows the warning dialog.
 *
 * @method
 *
 * @protected
 */
Keymaster.prototype.showWarningDialog = function () {
    this._warningDialog.style.display = "block";
};

/**
 * Hides the warning dialog.
 *
 * @method
 *
 * @protected
 */
Keymaster.prototype.hideWarningDialog = function () {
    this._warningDialog.style.display = "none";
};

/**
 * Requests the remaining session time from the server.
 *
 * @method
 *
 * @protected
 */
Keymaster.prototype.requestRemainingTime = function () {
    var request = new XMLHttpRequest();
    var url = "./?&keymaster_time";

    request.open("GET", url);
    // The following should simply use Function.prototype.bind() or a fallback,
    // but IE 8 doesn't set window.event, so we'd loose the reference to
    // the request object.
    request.onreadystatechange = (function (that) {
        return function () {
            that.receiveRemainingTime(request);
        };
    }(this));
    request.send(null);
};

/**
 * Handles responses for the requested session time.
 *
 * @method
 *
 * @param {XMLHttpRequest} request A request object.
 *
 * @protected
 */
Keymaster.prototype.receiveRemainingTime = function (request) {
    if (request.readyState == 4 && request.status == 200) {
        var remaining = +request.responseText;
        if (remaining != remaining) {
            clearInterval(this._timer);
            if (confirm(Keymaster.l10n.error)) {
                open(url);
            }
        } else if (remaining < 0) {
            location.href = location.href;
        } else if (remaining == 0) {
            location.href = "./?&logout";
        } else if (remaining < Keymaster.config.warn) {
            this.updateWarning(remaining);
            this.showWarningDialog();
        } else {
            this.hideWarningDialog();
        }
    }
};

/**
 * Resets the session timeout.
 *
 * @method
 *
 * @protected
 */
Keymaster.prototype.resetSession = function () {
    var request = new XMLHttpRequest();

    request.open("POST", "./");
    request.setRequestHeader("Content-Type",
            "application/x-www-form-urlencoded");
    request.send("keymaster_reset");
    this.hideWarningDialog();
};

/**
 * Updates the warning message.
 *
 * @method
 *
 * @param {Number} seconds The remaining time in seconds.
 *
 * @protected
 */
Keymaster.prototype.updateWarning = function (seconds) {
    var min = Math.ceil(seconds / 60);

    if (min == 1) {
        text = Keymaster.l10n.warning_singular;
    } else if (min >= 2 && min <= 4) {
        text = Keymaster.l10n.warning_paucal;
    } else {
        text = Keymaster.l10n.warning_plural;
    }
    this._warningText.nodeValue = text.replace(/{{{TIME}}}/, min);
};

/**
 * Creates a new function bound to this. Provides a workaround for possibly
 * missing Function.prototype.bind().
 *
 * @method
 *
 * @param {Function} func A function to bind.
 *
 * @returns {Function}
 *
 * @protected
 */
Keymaster.prototype.bind = function (func) {
    var that = this;

    return function () {
        func.apply(that, arguments);
    };
};

/*
 * Register Keymaster instantiation on load.
 */
if (typeof addEventListener != "undefined") {
    addEventListener("load", function () {
        new Keymaster();
    });
} else if (typeof attachEvent != "undefined") {
    attachEvent("onload", function () {
        new Keymaster();
    });
}
