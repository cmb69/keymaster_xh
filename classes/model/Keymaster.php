<?php

/**
 * Copyright (C) 2013-2023 Christoph M. Becker
 *
 * This file is part of Keymaster_XH.
 *
 * Keymaster_XH is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * Keymaster_XH is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with Keymaster_XH.  If not, see <http://www.gnu.org/licenses/>.
 */

namespace Keymaster\Model;

use Plib\Document;
use Plib\DocumentStore;

final class Keymaster implements Document
{
    /** @var string */
    private $key;

    /** @var int */
    private $timestamp;

    public static function updateIn(DocumentStore $store): self
    {
        $that = $store->update("key", self::class);
        assert($that instanceof self);
        return $that;
    }

    public static function fromString(string $contents, string $key): self
    {
        $record = explode(",", rtrim($contents), 2);
        return new self($record[0], count($record) === 2 ? (int) $record[1] : 0);
    }

    public function __construct(string $key, int $timestamp)
    {
        $this->key = $key;
        $this->timestamp = $timestamp;
    }

    public function toString(): string
    {
        return $this->key . "," . $this->timestamp;
    }

    private function hasKey(): bool
    {
        return $this->key !== "";
    }

    public function acceptKey(string $key, int $now, int $duration): bool
    {
        if ($key === $this->key) {
            $this->timestamp = $now;
            return true;
        }
        if ($this->isFree($now, $duration)) {
            $this->key = $key;
            $this->timestamp = $now;
            return true;
        }
        return false;
    }

    public function grantKey(callable $genKey, int $now, int $duration): ?string
    {
        if (!$this->isFree($now, $duration)) {
            return null;
        }
        $this->key = $genKey();
        return $this->key;
    }

    public function revokeKey(string $key): void
    {
        if ($this->key === $key) {
            $this->key = "";
            $this->timestamp = 0;
        }
    }

    public function isFree(int $now, int $duration): bool
    {
        return !$this->hasKey() || $now - $this->timestamp > $duration;
    }

    public function set(string $key, int $timestamp): void
    {
        $this->key = $key;
        $this->timestamp = $timestamp;
    }
}
