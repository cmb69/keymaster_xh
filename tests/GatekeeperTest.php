<?php

/**
 * Copyright (C) 2023 Christoph M. Becker
 *
 * This file is part of Keymaster_XH.
 *
 * Keymaster_XH is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * Keymaster_XH is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with Keymaster_XH.  If not, see <http://www.gnu.org/licenses/>.
 */

namespace Keymaster;

use ApprovalTests\Approvals;
use Keymaster\Model\Keymaster;
use org\bovigo\vfs\vfsStream;
use PHPUnit\Framework\MockObject\Stub;
use PHPUnit\Framework\TestCase;
use Plib\DocumentStore;
use Plib\FakeRequest;
use Plib\Random;
use Plib\View;

class GatekeeperTest extends TestCase
{
    /** @var DocumentStore */
    private $store;

    /** @var Random&Stub */
    private $random;

    /** @var View */
    private $view;

    public function setUp(): void
    {
        vfsStream::setup("root");
        $this->store = new DocumentStore(vfsStream::url("root/"));
        $keymaster = Keymaster::updateIn($this->store);
        $keymaster->set("12345", strtotime("2025-05-03T16:14:20+00:00"));
        $this->store->commit();
        $this->random = $this->createStub(Random::class);
        $this->random->method("bytes")->willReturn("123456789ABCDEF");
        $this->view = new View("./templates/", XH_includeVar("./languages/en.php", "plugin_tx")["keymaster"]);
    }

    private function sut(): Gatekeeper
    {
        return new Gatekeeper(
            "./plugins/keymaster/",
            XH_includeVar("./config/config.php", "plugin_cf")["keymaster"],
            $this->store,
            $this->random,
            $this->view
        );
    }

    public function testDoesNothingIfNothingNeedsToBeDone(): void
    {
        $request = new FakeRequest();
        $response = $this->sut()($request);
        $this->assertEquals("", $response->output());
    }

    public function testRevokesKeyAfterLogout(): void
    {
        $request = new FakeRequest([
            "url" => "http://example.com/?&logout",
            "cookie" => ["keymaster_key" => "12345"],
        ]);
        $response = $this->sut()($request);
        $this->assertSame("", $response->output());
    }

    public function testReportsFailureToRevokeKeyAfterLogout(): void
    {
        vfsStream::setQuota(0);
        $request = new FakeRequest([
            "url" => "http://example.com/?&logout",
            "cookie" => ["keymaster_key" => "12345"],
        ]);
        $response = $this->sut()($request);
        $this->assertStringContainsString("The key cannot be saved!", $response->output());
    }

    public function testAcceptsValidKey(): void
    {
        $request = new FakeRequest(["admin" => true, "cookie" => ["keymaster_key" => "12345"]]);
        $response = $this->sut()($request);
        $this->assertEmpty($response->output());
    }

    public function testReportsFailureToAcceptKey(): void
    {
        vfsStream::setQuota(0);
        $request = new FakeRequest(["admin" => true, "cookie" => ["keymaster_key" => "12345"]]);
        $response = $this->sut()($request);
        $this->assertStringContainsString("The key cannot be saved!", $response->output());
    }

    public function testGrantsKey(): void
    {
        $request = new FakeRequest([
            "admin" => true,
            "time" => strtotime("2025-05-03T17:14:20+00:00"),
        ]);
        $response = $this->sut()($request);
        $this->assertSame(["keymaster_key", "64P36D1L6ORJGEA1891K8HA6", 0], $response->cookie());
    }

    public function testReportsFailureToGrantKey(): void
    {
        vfsStream::setQuota(0);
        $request = new FakeRequest([
            "admin" => true,
            "time" => strtotime("2025-05-03T17:14:20+00:00"),
        ]);
        $response = $this->sut()($request);
        $this->assertStringContainsString("The key cannot be saved!", $response->output());
    }

    public function testRendersLockScreen(): void
    {
        $request = new FakeRequest([
            "admin" => true,
            "cookie" => ["keymaster_key" => "12346"],
            "time" => strtotime("2025-05-03T16:14:24+00:00"),
        ]);
        $response = $this->sut()($request);
        $this->assertSame(409, $response->status());
        Approvals::verifyHtml($response->output());
    }
}
