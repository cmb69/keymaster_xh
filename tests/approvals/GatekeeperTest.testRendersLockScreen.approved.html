<!DOCTYPE html>
<html>
  <head>
    <meta http-equiv="refresh" content="300">
    <link href="/plugins/keymaster/css/stylesheet.css" rel="stylesheet">
  </head>
  <body class="keymaster_lock_screen">
    <div class="keymaster_lock_message">
      <p>Somebody else is currently editing the site. Logout to view the site right away, or wait 5 minutes for an automatic retry to access the site as admin.</p>
      <form method="post" action="/?&amp;logout" class="keymaster_buttons">
        <button>Logout</button>
      </form>
    </div>
  </body>
</html>
